nativelib/src/main/resources/gc/commix/Marker.c

```diff
@@ -0,0 +1,380 @@
+#include <stdio.h>
+#include <setjmp.h>
+#include "Marker.h"
+#include "Object.h"
+#include "Log.h"
+#include "State.h"
+#include "headers/ObjectHeader.h"
+#include "datastructures/GreyPacket.h"
+#include "GCThread.h"
+#include <sched.h>
+
+extern word_t *__modules;
+extern int __modules_size;
+extern word_t **__stack_bottom;
+
+#define LAST_FIELD_OFFSET -1
+
+// Marking is done using grey packets. A grey packet is a fixes size list that
+// contains pointers to objects for marking.
+//
+// Each marker has a grey packet with references to check ("in" packet).
+// When it finds a new unmarked object the marker puts a pointer to
+// it in the "out" packet. When the "in" packet is empty it gets
+// another from the full packet list and returns the empty one to the empty
+// packet list. Similarly, when the "out" packet get full, marker gets another
+// empty packet and pushes the full one on the full packet list.
+//
+// Marking is done when all the packets are empty and in the empty packet list.
```

Comment:
Needs more details. Especially around:
1. Work balancing: array splitting, packet splitting (including explanation of the constants).
2. Synchronization: only on taking/returning packets from the pool, but need to mention that marking naturally doesn't require fine-grain synchronization.

Reply:
Also just a brief mention of how roots are handled and how/when are threads spin up/down would be good as well.